src/crimson/osd/osd_operations/client_request.cc

```diff
@@ -347,7 +349,8 @@ bool ClientRequest::is_misdirected(const PG& pg) const
 
 void ClientRequest::put_historic() const
 {
-  osd.get_shard_services().get_registry().put_historic(*this);
+  ceph_assert(put_historic_shard_services);
```

Comment:
Can this assert be ever failed? `put_historic_shard_services` is initialized to `nullptr` at the definition site but the ctor sets it to address of one of its parameters. Do we expect introduction of a second constructor?

Reply:
We can demonstrate that statically now, but I'd really like to leave the assert.  I take it that it should be ceph_assert_always()?